File: zigzagLevelOrder.py
from collections import deque
class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right

class Solution:

# Approach 1- BFS
    def zigzagLevelOrder(self,root): # TC O(n) // SC O(n)
        ans=[]
        if not root: return ans
        q=deque()
        q.append(root)
        level=0
        while q:
            lst=[]
            for i in range(len(q)):
                currNode=q.popleft()
                lst.append(currNode.val)
                if currNode.left: q.append(currNode.left)
                if currNode.right: q.append(currNode.right)
            if level%2:
                lst.reverse()
            level+=1
            ans.append(lst)
        
        return ans

# Approach 2- BFS 
    def zigzagLevelOrder1(self, root): # TC O(n) // SC O(n)
        ret = []
        level_list = deque()
        if root is None:
            return []
        # start with the level 0 with a delimiter
        node_queue = deque([root, None])
        is_order_left = True

        while len(node_queue) > 0:
            curr_node = node_queue.popleft()

            if curr_node:
                if is_order_left:
                    level_list.append(curr_node.val)
                else:
                    level_list.appendleft(curr_node.val)

                if curr_node.left:
                    node_queue.append(curr_node.left)
                if curr_node.right:
                    node_queue.append(curr_node.right)
            else:
                # we finish one level
                ret.append(level_list)
                # add a delimiter to mark the level
                if len(node_queue) > 0:
                    node_queue.append(None)

                # prepare for the next level
                level_list = deque()
                is_order_left = not is_order_left

        return ret


# Approach 3- DFS 
    def zigzagLevelOrder2(self, root): # TC O(n) // SC O(n)
        if root is None:
            return []

        results = []
        def dfs(node, level):
            if level >= len(results):
                results.append(deque([node.val]))
            else:
                if level % 2 == 0:
                    results[level].append(node.val)
                else:
                    results[level].appendleft(node.val)

            for next_node in [node.left, node.right]:
                if next_node is not None:
                    dfs(next_node, level+1)

        # normal level order traversal with DFS
        dfs(root, 0)

        return results


root=TreeNode(3)
root.left=TreeNode(9)
root.right=TreeNode(20)
root.right.left=TreeNode(15)
root.right.right=TreeNode(7)
s=Solution()
print(s.zigzagLevelOrder(root))
print(s.zigzagLevelOrder1(root))
print(s.zigzagLevelOrder2(root))

File: zigzagLevelOrder.java
import java.util.*;
// Approach 1- BFS 
class Solution{ // TC O(n) SC O(n)
	public List<List<Integer>> zigzagLevelOrder(Node root){
		List<List<Integer>> ans=new ArrayList<>();
		if(root==null) return ans;
		Queue<Node> queue=new LinkedList<>();
		queue.offer(root);
		int level=0;
		while(!queue.isEmpty()){
			int size=queue.size();
			List<Integer> list=new ArrayList<>();
			for(int i=0; i<size; i++){
				Node currNode=queue.poll();
				list.add(currNode.val);
				if(currNode.left!=null) queue.add(currNode.left);
				if(currNode.right!=null) queue.add(currNode.right);
			}
			if(level%2==1){
				Collections.reverse(list);
			}
			ans.add(new ArrayList<>(list));
			level++;
		}
		return ans;
	}
}

// Approach 2 - BFS 
class Solution1{ // TC O(n) SC O(n)
  public List<List<Integer>> zigzagLevelOrder(Node root) {
    if (root == null) {
      return new ArrayList<List<Integer>>();
    }

    List<List<Integer>> results = new ArrayList<List<Integer>>();

    // add the root element with a delimiter to kick off the BFS loop
    LinkedList<Node> node_queue = new LinkedList<Node>();
    node_queue.addLast(root);
    node_queue.addLast(null);

    LinkedList<Integer> level_list = new LinkedList<Integer>();
    boolean is_order_left = true;

    while (node_queue.size() > 0) {
      Node curr_node = node_queue.pollFirst();
      if (curr_node != null) {
        if (is_order_left)
          level_list.addLast(curr_node.val);
        else
          level_list.addFirst(curr_node.val);

        if (curr_node.left != null)
          node_queue.addLast(curr_node.left);
        if (curr_node.right != null)
          node_queue.addLast(curr_node.right);

      } else {
        // we finish the scan of one level
        results.add(level_list);
        level_list = new LinkedList<Integer>();
        // prepare for the next level
        if (node_queue.size() > 0)
          node_queue.addLast(null);
        is_order_left = !is_order_left;
      }
    }
    return results;
  }
}

// Approach 3- DFS 
class Solution1{ // TC O(n) SC O(n)
  protected void DFS(Node node, int level, List<List<Integer>> results) {
    if (level >= results.size()) {
      LinkedList<Integer> newLevel = new LinkedList<Integer>();
      newLevel.add(node.val);
      results.add(newLevel);
    } else {
      if (level % 2 == 0)
        results.get(level).add(node.val);
      else
        results.get(level).add(0, node.val);
    }

    if (node.left != null) DFS(node.left, level + 1, results);
    if (node.right != null) DFS(node.right, level + 1, results);
  }

  public List<List<Integer>> zigzagLevelOrder(Node root) {
    if (root == null) {
      return new ArrayList<List<Integer>>();
    }
    List<List<Integer>> results = new ArrayList<List<Integer>>();
    DFS(root, 0, results);
    return results;
  }
}

class zigzagLevelOrder{
	public static void main(String[] args) {
		BinaryTree bt= new BinaryTree();
		bt.root=new Node(3);
		bt.root.left=new Node(9);
		bt.root.right=new Node(20);
		bt.root.right.left=new Node(15);
		bt.root.right.right=new Node(7);
		Solution s= new Solution();
		System.out.println(s.zigzagLevelOrder(bt.root));
	}
}